junit-jupiter-api/src/main/java/org/junit/jupiter/api/AssertThrows.java

```diff
@@ -38,8 +41,37 @@
 				throw new AssertionFailedError(message, actualException);
 			}
 		}
-		throw new AssertionFailedError(
-			String.format("Expected %s to be thrown, but nothing was thrown.", getCanonicalName(expectedType)));
+		throw new AssertionFailedError(getFailureMessage(expectedType));
+	}
+
+	static <T extends Throwable> T assertThrows(Class<T> expectedType, Executable executable, String message) {
+		return assertThrows(expectedType, executable, () -> message);
 	}
 
+	@SuppressWarnings("unchecked")
+	static <T extends Throwable> T assertThrows(Class<T> expectedType, Executable executable,
+			Supplier<String> messageSupplier) {
+		try {
+			executable.execute();
+		}
+		catch (Throwable actualException) {
+			if (expectedType.isInstance(actualException)) {
+				return (T) actualException;
+			}
+			else {
+				throw new AssertionFailedError(getFailureWIthSuppliedMessage(expectedType, messageSupplier),
+					actualException);
+			}
+		}
+		throw new AssertionFailedError(getFailureWIthSuppliedMessage(expectedType, messageSupplier));
+	}
+
+	private static <T extends Throwable> String getFailureWIthSuppliedMessage(Class<T> expectedType,
+			Supplier<String> messageSupplier) {
+		return nullSafeGet(messageSupplier) + " " + getFailureMessage(expectedType);
```

Comment:
For consistency with the rest of the assertions, please use `org.junit.jupiter.api.AssertionUtils.buildPrefix(String)` instead of concatenating the two strings with a space.

Consult existing assertions for inspiration. 😉